implement Concurrent GeneratorFilter technique

PhyM73 opened the pull request:
#### PR description:

The concurrent version of `GeneratorFilter` is implemented for GEN workflows. This global module mimics the `ConcurrentHadronizerFilter` and replicates the Generator for each stream in order to run them concurrently. This only works for thread-friendly hadronizers, decayers, and filters.

- Nothing changed in output; expects improved CPU time.
- None of other PRs or externals it depends upon.
- Add a concurrent module of Pythia8 generator, `Pythia8ConcurrentGeneratorFilter`.
- Provide a unit test in `GeneratorInterface/Pythia8Interface/test/BuildFile.xml` and a test file as `GeneratorInterface/Core/test/ConcurrentGeneratorFilter_WZ_TuneCP5_13TeV-pythia8_cfg.py`.
- [#25090(comment)](https://github.com/cms-sw/cmssw/issues/25090#issuecomment-819840962).

#### PR validation:

Tested on [BTV-RunIISummer20UL17GEN-00002](https://cms-pdmv.cern.ch/mcm/requests?prepid=BTV-RunIISummer20UL17GEN-00002&page=0&shown=127) with 2, 4, and 8 threads and with 1 and 2 ConcurrentLuminosityBlocks respectively. All of them succeed.

PhyM73 committed:
Add ConcurrentGeneratorFilter and a test file.

Implement Concurrent GeneratorFilter technique, which can replicate the Generator for each stream in order to run them concurrently. Add a concurrent module of pythia8 generator and a python _cfg file for test.
Clean the redundance in `GeneratorInterface/Pythia8Interface/plugins/BuildFile.xml`.

PhyM73 committed:
Add a unit test for ConcurrentGeneratorFilter

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33883/22920

- This PR adds an extra 28KB to repository

- Found files with invalid states:
  - GeneratorInterface/Core/test/ConcurrentGeneratorFilter_WZ_TuneCP5_13TeV-pythia8_cfg.py:
    - Added: cd2ff5af85d377009ce1f5693262a00106843c7b
    - Deleted: fc9f8db7009b25224381a82dc50e4f7e498c6e67

cmsbuild commented:
A new Pull Request was created by  @PhyM73 (Frank M) for master.

It involves the following packages:

GeneratorInterface/Core
GeneratorInterface/Pythia8Interface


@SiewYan, @mkirsano, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @alberto-sanchez can you please review it and eventually sign? Thanks.
@alberto-sanchez, @mkirsano this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

SiewYan commented:
please test

PhyM73 commented:
@makortel  @Dr15Jones FYI. : ）

Dr15Jones commented:
@PhyM73  Thanks!

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bb2040/15391/summary.html
**COMMIT**: fc9f8db7009b25224381a82dc50e4f7e498c6e67
**CMSSW**: CMSSW_12_0_X_2021-05-28-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/33883/15391/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2650486
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2650457
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 36 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 155 log files, 37 edm output root files, 37 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-05-28-1100+bb2040/43031/triggerResults): no differences found

PhyM73 committed:
Fix the shell script to test the Py8ConGenFilter.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33883/22932

- This PR adds an extra 28KB to repository

- Found files with invalid states:
  - GeneratorInterface/Core/test/ConcurrentGeneratorFilter_WZ_TuneCP5_13TeV-pythia8_cfg.py:
    - Added: cd2ff5af85d377009ce1f5693262a00106843c7b
    - Deleted: fc9f8db7009b25224381a82dc50e4f7e498c6e67

cmsbuild commented:
Pull request #33883 was updated. @SiewYan, @mkirsano, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @alberto-sanchez can you please check and sign again.

SiewYan commented:
pleas test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bb2040/15403/summary.html
**COMMIT**: 91e9bdd94d02fe4231a9d6b2895e6afc3e03a837
**CMSSW**: CMSSW_12_0_X_2021-05-28-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/33883/15403/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 10 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2650486
 - DQMHistoTests: Total failures: 18
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2650445
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 36 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 155 log files, 37 edm output root files, 37 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-05-28-2300+bb2040/43042/triggerResults): no differences found

SiewYan commented:
The unit test is successful. Its look all good to me.

agrohsje commented on GeneratorInterface/Pythia8Interface/test/test_Pythia8ConcurrentGeneratorFilter_WZ_TuneCP5_13TeV-pythia8.sh:
```diff
@@ -0,0 +1,13 @@
+scram b
+cd ../..
+
+cmsDriver.py Configuration/GenProduction/python/BTV-RunIISummer20UL17GEN-00002-fragment.py --python_filename test_BTV-RunIISummer20UL17GEN-00002_1_cfg.py --eventcontent RAWSIM --customise Configuration/DataProcessing/Utils.addMonitoring --datatier GEN --fileout file:test_BTV-RunIISummer20UL17GEN-00002.root --conditions auto:run2_mc --beamspot Realistic25ns13TeVEarly2017Collision --customise_comma ...
```
Shouldn't the cmsDriver be in sync with the above script? There 100 events are produced and nThreads 4. Should we include a transition in lumisection as this is know to be problematic. I.e. produce a total of 100 events and 10 per lumi block? @colizz @qliphy @PhyM73 what about adding concurrent workflows to relval?

PhyM73 replied:
Thanks for pointing this out. I didn't require this `cmsDriver` to run too many events since I thought the _unittest_ might need to be a quick test. But I am unsure about this, maybe it could be a more comprehensive test? If so, then running 100 events with 10 events per lumi block on 4 threads or even more would be preferred of course. : )

I did test 10000 events with this module, and I haven't seen any problem with transitions yet. Another way to do local tests is using the `_cfg.py` file I provided, i.e., 

```bash
cd CMSSW_12_0_X/src
cmsenv
cmsRun ${CMSSW_BASE}/src/GeneratorInterface/Core/test/ConcurrentGeneratorFilter_WZ_TuneCP5_13TeV-pythia8_cfg.py
```

This way would be more convenient since its parameters, e.g., _events number_, _lumiblocks_, and _threads_, could be modified easily, and it does not need to build each time. However, it may only work under CMSSW_12_0_X.

PhyM73 commented:
I have two questions about the module. 

### About [Random Number Generator Service](https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideEDMRandomNumberGeneratorService) 

On this twiki page, [SWGuideEDMRandomNumberGeneratorService](https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideEDMRandomNumberGeneratorService#Ensuring_Events_are_Generated_wi), it is said that "_before every beginLuminosityBlock, the state of the (random) engines used in beginLuminosityBlock is restored from the cache to their original state._"  The purpose of the restore seems to be supporting replay. However, in the implementation of `ConcurrentHadronizerFilter`, the random number generator seed is advanced before each new lumi block begins. 
https://github.com/cms-sw/cmssw/blob/7cdba48db0e4ebf3638b32f09af699eda400fb70/GeneratorInterface/Core/interface/ConcurrentHadronizerFilter.h#L523-L528
I am not sure which way to follow. For now, I simply imitated `ConcurrentHadronizerFilter` for `ConcurrentGeneratorFilter`, i.e., advancing the seed at each lumi block. 


### About [Global module](https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkGlobalModuleInterface)

This [twiki page](https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkGlobalModuleInterface#edm_StreamCacheT) said that the `edm::SteamCache<T>` extension of the `edm::global` module allows it to use thread-unsafe classes as below: 
> Since a Stream processed data serially, the data held by the class `T` does not have to be thread-safe since the framework guarantees that the methods which take an `edm::StreamID` as an argument are never called simultaneously using the same `edm::StreamID` instance.

It also provides an example of using a non-thread safe helper class. I wonder if this framework could let us use some thread-unsafe decayer, such as EvtGen, within the `ConcurrentHadronizerFilter` or `ConcurrentGeneratorFilter`? I saw some static variables in the EvtGen interface, not sure if the `edm::global` module could protect them. 
https://github.com/cms-sw/cmssw/blob/7cdba48db0e4ebf3638b32f09af699eda400fb70/GeneratorInterface/EvtGenInterface/plugins/EvtGen/EvtGenInterface.h#L71

I want to compare `GeneratorFilter` with `ConcurrentGeneratorFilter` using Pythia8 + EvtGen directly to see if there are thread problems such as data races. This takes me to the replay and random number service. Thus the first question above comes out.

Please give me some advice, thanks!
@Dr15Jones @makortel @qliphy @wddgit  @SiewYan @agrohsje @colizz

Dr15Jones commented:
> It also provides an example of using a non-thread safe helper class. I wonder if this framework could let us use some thread-unsafe decayer, such as EvtGen, within the ConcurrentHadronizerFilter or ConcurrentGeneratorFilter? I saw some static variables in the EvtGen interface, not sure if the edm::global module could protect them.

There is a difference between non _thread-safe_ and being what I call _thread hostile_. Non _thread-safe_ is an object which can't have multiple threads modifying it at the same time but mulitple threads can each modify different instances of that class safely (such as a `std::vector<>`). A _thread hostile_ code is one that even different instances of a class can not be modified by different threads.

EvtGen is _thread hostile_ as it uses global state (e.g. FORTRAN common blocks) in is implementation. Therefore it can not be safely accessed using the `StreamCache` system. If if could, we would have just make the generators into `stream::` modules.

wddgit commented:
About the Random number question. It is possible I am missing something or have forgotten something, I've only thought about this for about 1 minute since I read it and it has been at least a couple years since I looked into the random code ... 

But it looks to me like that line of code running in global end luminosity block produce will have no effect other than to waste a small number of CPU cycles (negligibly small). The random engine state will get overwritten before begin lumi and the fact that an extra random number was generated has no effect on the next begin lumi or the event engines (which are different engines anyway). You could remove it and see if that has any effect at all, just print out the first random number at begin lumi...

PhyM73 commented:
@wddgit It seems those lines of code have no effects. I print out the first random number at each begin lumi, and they are all the same. BTW, the random engine is cloned at each streamBeginLuminosityBlock, so they should be equivalent. : )

https://github.com/cms-sw/cmssw/blob/7cdba48db0e4ebf3638b32f09af699eda400fb70/GeneratorInterface/Core/interface/ConcurrentHadronizerFilter.h#L381-L401

@Dr15Jones I see. So the only way to run Pythia8 + EvtGen concurrently now is using `ExternalGeneratorFilter`.

PhyM73 committed:
Remove the extra random flat, extend unittest.
Remove the extra random number advance in global end lumi. Extend the
unit test to 50 events.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33883/23108

- This PR adds an extra 32KB to repository

- Found files with invalid states:
  - GeneratorInterface/Core/test/ConcurrentGeneratorFilter_WZ_TuneCP5_13TeV-pythia8_cfg.py:
    - Added: cd2ff5af85d377009ce1f5693262a00106843c7b
    - Deleted: fc9f8db7009b25224381a82dc50e4f7e498c6e67

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33883/23108/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33883/23108/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

PhyM73 committed:
Code Format

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33883/23109

- This PR adds an extra 32KB to repository

- Found files with invalid states:
  - GeneratorInterface/Core/test/ConcurrentGeneratorFilter_WZ_TuneCP5_13TeV-pythia8_cfg.py:
    - Added: cd2ff5af85d377009ce1f5693262a00106843c7b
    - Deleted: fc9f8db7009b25224381a82dc50e4f7e498c6e67

cmsbuild commented:
Pull request #33883 was updated. @SiewYan, @mkirsano, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @alberto-sanchez can you please check and sign again.

SiewYan commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bb2040/15717/summary.html
**COMMIT**: 6e43c45cb258ec18771b97094b775789a413d504
**CMSSW**: CMSSW_12_0_X_2021-06-06-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/33883/15717/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2648335
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2648312
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 36 files compared)
 - Checked 155 log files, 37 edm output root files, 37 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-06-06-2300+bb2040/43265/triggerResults): no differences found

agrohsje commented:
+generators

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1
